apps/prairielearn/src/lib/grading.js

```diff
@@ -155,158 +113,103 @@ export function saveSubmission(submission, variant, question, variant_course, ca
-            broken: hasFatalIssue,
-          };
-        } else {
-          // for External grading we don't do anything
-          courseIssues = [];
-          data = {};
-        }
+  const question_course = await getQuestionCourse(question, variant_course);
 
-        const studentMessage = 'Error grading submission';
-        const courseData = { variant, question, submission, course: variant_course };
-        await writeCourseIssues(
-          courseIssues,
-          variant,
-          submission.auth_user_id,
-          studentMessage,
-          courseData,
-        );
-      },
-      (callback) => {
-        if (question.grading_method === 'External') {
-          // We haven't actually graded this question yet - don't attempt
-          // to update the grading job or submission.
-          return callback(null);
-        }
-
-        const params = [
-          grading_job.id,
-          // `received_time` and `start_time` were already set when the
-          // grading job was inserted, so they'll remain unchanged.
-          // `finish_time` will be set to `now()` by this sproc.
-          null, // received_time
-          null, // start_time
-          null, // finish_time
-          data.submitted_answer,
-          data.format_errors,
-          data.gradable,
-          data.broken,
-          data.params,
-          data.true_answer,
-          data.feedback,
-          data.partial_scores,
-          data.score,
-          data.v2_score,
-        ];
-        sqldb.callOneRow('grading_jobs_update_after_grading', params, (err, result) => {
-          if (ERR(err, callback)) return;
+  const submissionResult = await sqldb.callZeroOrOneRowAsync(
+    'variants_select_submission_for_grading',
+    [variant.id, check_submission_id],
+  );
+  if (submissionResult.rowCount == null || submissionResult.rowCount === 0) return;
+  const submission = submissionResult.rows[0];
+
+  if (!overrideGradeRateCheck) {
+    const resultNextAllowed = await sqldb.callOneRowAsync('instance_questions_next_allowed_grade', [
+      variant.instance_question_id,
+    ]);
+    if (resultNextAllowed.rows[0].allow_grade_left_ms > 0) return;
+  }
 
-          // If the submission was marked invalid during grading the grading
-          // job will be marked ungradable and we should bail here to prevent
-          // LTI updates.
-          grading_job = result.rows[0];
-          if (!grading_job.gradable) return callback(new NoSubmissionError());
+  const resultGradingJob = await sqldb.callOneRowAsync('grading_jobs_insert', [
+    submission.id,
+    authn_user_id,
+  ]);
+  const grading_job = resultGradingJob.rows[0];
+
+  if (question.grading_method === 'External') {
+    // For external grading we just need to trigger the grading job to start.
+    // We haven't actually graded this question yet - don't attempt
+    // to update the grading job or submission.
+    await externalGrader.beginGradingJob(grading_job.id);
+  } else {
```

Comment:
This was originally in the callback of async.series, to be executed at the end if the data was not broken, but data is never updated for external grader, so running it right after inserting the grading job gives the same result.